### PR TITLE
 feat: Add URI validator and set default value for resource path pattern in the Blazor recipe

### PR DESCRIPTION
**Pull request by 96malhar**
*Issue #, if available:*
* DOTNET-5805
* DOTNET-5806

*Description of changes:*
* Add new `UriValidator` class
* Enforces the `UriValidator` and `RegexValidator` on `BackendApi.Uri`.
* Set default value for `BackendApi.ResourcePathPattern` and impose the `RegexValidator` on it.

image
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by ashovlin — COMMENTED** [review]
While we're in this area - we're not enforcing the requirement on `Uri` which is failing during CDK deploy if left blank. 

```
Unhandled exception. System.UriFormatException: Invalid URI: The URI is empty.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString)
   at BlazorWasm50.Recipe.ConfigureCloudFrontDistribution(Configuration settings) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Generated\Recipe.cs:line 167
   at BlazorWasm50.Recipe..ctor(Construct scope, IRecipeProps`1 props) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Generated\Recipe.cs:line 48
   at BlazorWasm50.AppStack..ctor(Construct scope, IDeployToolStackProps`1 props) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\AppStack.cs:line 31
   at BlazorWasm50.Program.Main(String[] args) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Program.cs:line 29

Subprocess exited with error 3762504530
We were unable to delete the temporary project that was created for this deployment. 
```

Could we add that why we're here? If not, is there a separate issue tracking that?

**Comment by philasmar** [comment]
> While we're in this area - we're not enforcing the requirement on `Uri` which is failing during CDK deploy if left blank.
> 
> I'm allowed to publish this: image
> 
> But it results in:
> 
> ```
> Unhandled exception. System.UriFormatException: Invalid URI: The URI is empty.
>    at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
>    at System.Uri..ctor(String uriString)
>    at BlazorWasm50.Recipe.ConfigureCloudFrontDistribution(Configuration settings) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Generated\Recipe.cs:line 167
>    at BlazorWasm50.Recipe..ctor(Construct scope, IRecipeProps`1 props) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Generated\Recipe.cs:line 48
>    at BlazorWasm50.AppStack..ctor(Construct scope, IDeployToolStackProps`1 props) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\AppStack.cs:line 31
>    at BlazorWasm50.Program.Main(String[] args) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Program.cs:line 29
> 
> Subprocess exited with error 3762504530
> We were unable to delete the temporary project that was created for this deployment. 
> ```
> 
> Could we add that why we're here? If not, is there a separate issue tracking that?

The story that i am working on will help in this area. So when you hit enable Backend rest api, that'll run the validators of the dependent settings. Only issue here is that the URI does not have a regex validator. So I think part of this story, Malhar should add a regex validator, which will be automatically executed once I finish my feature.

**Review by philasmar — CHANGES_REQUESTED** [review]
https://github.com/aws/aws-dotnet-deploy/pull/572#issuecomment-1137800290

**Comment by 96malhar** [comment]
> While we're in this area - we're not enforcing the requirement on `Uri` which is failing during CDK deploy if left blank.
> 
> I'm allowed to publish this: image
> 
> But it results in:
> 
> ```
> Unhandled exception. System.UriFormatException: Invalid URI: The URI is empty.
>    at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
>    at System.Uri..ctor(String uriString)
>    at BlazorWasm50.Recipe.ConfigureCloudFrontDistribution(Configuration settings) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Generated\Recipe.cs:line 167
>    at BlazorWasm50.Recipe..ctor(Construct scope, IRecipeProps`1 props) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Generated\Recipe.cs:line 48
>    at BlazorWasm50.AppStack..ctor(Construct scope, IDeployToolStackProps`1 props) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\AppStack.cs:line 31
>    at BlazorWasm50.Program.Main(String[] args) in C:\Users\shovlia\.aws-dotnet-deploy\Projects\cv24keup\Program.cs:line 29
> 
> Subprocess exited with error 3762504530
> We were unable to delete the temporary project that was created for this deployment. 
> ```
> 
> Could we add that why we're here? If not, is there a separate issue tracking that?

Added a default value for the URI field and also added the appropriate validators to the URI and the resource path pattern.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe [hunk 1]
```diff
@@ -166,29 +166,43 @@
                     "Name": "Uri",
                     "Description": "Uri to the backend rest api",
                     "Type": "String",
-                    "DefaultValue": "",
+                    "DefaultValue": "http://api-uri",
```
Comment (normj) [inline src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: I think we should have no default value for this field so the user has to enter a value. We want the toolkit to flag you must put a value here before publishing.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe (hunk 1 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: To Norm's point, even if a user tries to deploy with an invalid value, the validators will all be run before we kick off deployments which should catch this https://github.com/aws/aws-dotnet-deploy/blob/e2ad9b5a026869aa8854182dd5df1f4cedc6fa46/src/AWS.Deploy.CLI/Commands/DeployCommand.cs#L723

Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: In general, for this https://github.com/aws/aws-dotnet-deploy/blob/main/src/AWS.Deploy.CLI/Commands/DeployCommand.cs#L834, I vote to remove that check. any objections @normj ?

**Review by normj — APPROVED** [review]

File: src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe (hunk 1 above)
Comment (96malhar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: We can remove this check (If we decide to) as part of a separate PR.

Comment (96malhar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: Also, setting invalid default values can easily trip customers as they could be performing deployments via a config file and may not specify an explicit value for all option settings

**Review by philasmar — APPROVED** [review]

File: src/AWS.Deploy.Common/Recipes/Validation/OptionSettingItemValidators/UriValidator.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        public string ValidationFailedMessage { get; set; } = "{{URI}} is not a valid URI.";
+
+        public Task<ValidationResult> Validate(object input, Recommendation recommendation)
+        {
+            var uri = input?.ToString() ?? string.Empty;
+
+            if (string.IsNullOrEmpty(uri))
+            {
+                return ValidationResult.FailedAsync("Invalid URI. The URI cannot be null or empty");
```
Comment (ashovlin) [inline src/AWS.Deploy.Common/Recipes/Validation/OptionSettingItemValidators/UriValidator.cs:24]: Is true of all future `Uri` options? 

If not and we may conceivably add an option Uri option in the future, I'd rather use a `Required` validator here to enforce that a value is provided.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe (hunk 1 above)
Comment (ashovlin) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: But `"http://api-uri"` is likely invalid for the user's application that they're deploying, right? 

I don't think we should be defaulting a value just to get past client side validation, then deploy an application that likely won't work.

It feels like we have to have a concept/experience for _"this field is required even though there isn't a reasonable default value so you must provide one"_

File: src/AWS.Deploy.Common/Recipes/Validation/OptionSettingItemValidators/UriValidator.cs (hunk 2 above)
Comment (96malhar) [reply src/AWS.Deploy.Common/Recipes/Validation/OptionSettingItemValidators/UriValidator.cs:24]: I cannot foresee such a use case right now. But even if we encounter it in the future than it should be easy to remove the
`string.IsNullOrEmpty(uri)`  check. Also, this removal would not be a breaking change as we would be making the validator more permissive.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe (hunk 1 above)
Comment (96malhar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe:169]: This needs a larger discussion.

File: src/AWS.Deploy.Common/Recipes/Validation/OptionSettingItemValidators/UriValidator.cs (hunk 2 above)
Comment (ashovlin) [reply src/AWS.Deploy.Common/Recipes/Validation/OptionSettingItemValidators/UriValidator.cs:24]: What if a user is using a custom recipe with a Uri option? Could they modify the source of the validators?

In general I would prefer validators as self-contained and composable as possible.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe (hunk 1 above)
Comment (96malhar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/BlazorWasm.recipe]: After syncing up with @normj, removed the default value on `BackendApi.Uri` and also removed  https://github.com/aws/aws-dotnet-deploy/blob/main/src/AWS.Deploy.CLI/Commands/DeployCommand.cs#L834

**Review by philasmar — APPROVED** [review]

**Review by normj — APPROVED** [review]

**Pull request merged**